### ISSUE TITLE
从数据库里取sql表数据后，怎样通过自定义处理数据脚本处理后，在训练时shell文件中配置参数

### ISSUE THREAD

**Issue by onex7777** [modelscope/ms-swift#5611]
我首先写一个自处理数据的脚本， 然后通过sh文件启动它，但是它一直会自动下载文件，而不是使用我脚本处理注册的数据集，请问这个怎么配置？

odps_dataset.py  -----自定义数据处理脚本

 # Copyright (c) Alibaba, Inc. and its affiliates.
from typing import Any, Dict, Optional, List
import json
import os
from odps import ODPS
from odps.tunnel import TableTunnel
from odps import DataFrame

from swift.llm import DatasetMeta, ResponsePreprocessor, load_dataset, register_dataset

"""
{"messages": 
    [   
        {"role": "system", "content": "You are a helpful assistant."}, 
        {"role": "user", "content": "<image>帮我打开谷歌浏览器"}, 
        {"role": "assistant", "content": "Action: click(start_box='<|box_start|>(246,113)<|box_end|>')"}
    ], 
"images": ["/xxx/x.jpg"]}
"""

class ODPSMultiModalPreprocessor(ResponsePreprocessor):
    """
    从 ODPS 表加载 prompt + photo_url + label，并构造为多模态 messages 格式
    支持 photo_url 为空的情况
    """

    def __init__(self, odps_config_path: str = None):
        super().__init__()
        self.odps_config_path = odps_config_path or os.getenv('ODPS_CONFIG_PATH', 'script/odps_config.json')

    def load_from_odps(self) -> List[Dict[str, Any]]:
        """从 ODPS 表加载数据并构造 messages"""
        if not os.path.exists(self.odps_config_path):
            raise FileNotFoundError(f"ODPS config not found: {self.odps_config_path}")

        with open(self.odps_config_path, 'r') as f:
            config = json.load(f)

        # 初始化 ODPS 客户端
        odp = ODPS(
            access_id=config.get('access_id') or os.getenv('ODPS_ACCESS_ID'),
            secret_access_key=config.get('access_key') or os.getenv('ODPS_ACCESS_KEY'),
            project=config['project'],
            endpoint=config['endpoint']
        )
        try:
            table_name = os.getenv('TABLE_PATH', '')
            schema = os.getenv('SCHEMA', None)
            df = DataFrame(odp.get_table(name=table_name, project=config['project'], schema=schema)).to_pandas()
            reader = df.to_dict('records')
        except Exception as e:
            print("e")
        dataset = []
        for record in reader:
            try:
                prompt_text = str(record["prompt"] or "").strip()
                photo_url = str(record["photo_url"] or "").strip() if record["photo_url"] else None
                label = str(record["label"] or "").strip()

                if not prompt_text or not label:
                    continue  # 跳过空数据
                # 构造标准 messages
                messages = [
                    {"role": "user", "content": prompt_text},
                    {"role": "assistant", "content": label}
                ]
                sub_data = {"messages": messages}
                # 构造 content 列表（支持多模态）

                if photo_url:
                    sub_data["images"] = photo_url
                dataset.append(sub_data)
            except Exception as e:
                print(f"[Warning] Skip invalid record: {e}")
                continue

        return dataset
    def preprocess(self, row: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """
        已经是标准格式，直接返回
        这里可以做额外清洗（如长度过滤）
        """
        messages = row.get("messages")
        if not messages or len(messages) < 2:
            return None
        return row
# 实例化预处理器
preprocessor = ODPSMultiModalPreprocessor()

# 注册数据集
register_dataset(
    DatasetMeta(
        dataset_path='my_odps_dataset',  # 自定义 dataset ID
        preprocess_func=preprocessor,
        tags=["llm", "mllm"],  #
    )
)

# 调试：本地打印数据
if __name__ == '__main__':
    dataset = load_dataset(['my_odps_dataset'])[0]
    print(f'dataset: {dataset}')
    print(f'dataset[0]: {dataset[0]}')

下面是启动脚本的部分参数：

export TABLE_PATH=query_item_standard_label_data_2025_llm_27w_for_et_multi_feature_filtered_train
export SCHEMA=20250821_min   # 具体schema名或为None

PROMPT_TEMPLATE="qwen3"
MODEL_NAME="Qwen/Qwen3-8B"

INSTRUCTION="prompt"
RESPONSE="label"

# ✅ 使用平台导出的本地路径
args="--rlhf_type grpo \
    --generation_batch_size 8 \
    --attn_impl flash_attention_2 \
    --custom_register_path examples/custom/odps_dataset.py \
    --model $MODEL_NAME \
    --dataset my_odps_dataset \
    --eval_strategy epoch  \
    --output_dir=local/tmp/ckpt_save_path/ \
    --reward_funcs accuracy \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.55 \
    --vllm_max_model_len 2048 \
    --vllm_tensor_parallel_size 2 \
    --offload_optimizer true \
    --offload_model true \
    --train_type lora \
    --lora_rank 32 \
    --lora_alpha 64 \
    --lora_dropout 0.05 \
    --report_to=ml_tracker \
    --torch_dtype bfloat16 \
    --columns '{\"text1\": \"prompt\", \"text2\": \"\"}' \
    --split_dataset_ratio 0.01 \
    --max_completion_length 4096 \
    --num_train_epochs 3 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --learning_rate ${LR} \
    --gradient_accumulation_steps 4 \
    --save_total_limit 2 \
    --logging_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --temperature 1.0 \
    --top_p 0.9 \
    --top_k 50 \
    --deepspeed zero2 \
    --log_completions true \
    --num_iterations 1 \
    --beta 0"

报错log：
[INFO:swift] swift.__version__: 3.8.0.dev0
[INFO:swift] Downloading the dataset from ModelScope, dataset_id: my_odps_dataset
[ERROR:modelscope] Repo modelscope/my_odps_dataset not exists on either https://www.modelscope.cn or https://www.modelscope.ai
[ERROR:swift] Dataset my_odps_dataset load failed: subset_name=default,split=train with error: <Response [404]>
[ERROR:modelscope] Repo modelscope/my_odps_dataset not exists on either https://www.modelscope.cn or https://www.modelscope.ai
[ERROR:swift] Dataset my_odps_dataset load failed: subset_name=default,split=train with error: <Response [404]>
[ERROR:modelscope] Repo modelscope/my_odps_dataset not exists on either https://www.modelscope.cn or https://www.modelscope.ai

**Comment by Jintao-Huang** [comment]
my_odps_dataset是文件夹还是路径

如果是文件夹，你试试ms_dataset_id='xxx'

**Comment by onex7777** [comment]
不是文件夹 而是odps数据处理后，metadata类里面自己自定义的一个名字（就是在shell文件传入的一个odps＿dataset.py脚本自动化处理odps数据（在执行shell文件时自动化拉去数据并处理成模型可用的数据，不需要线下处理），想让这个处理完的数据可以直接用于训练或推理）



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年08月31日 23:35 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [modelscope/ms-swift] 从数据库里取sql表数据后，怎样通过自定义处理数据脚本处理后，在训练时shell文件中配置参数 (Issue #5611) |
Jintao-Huang left a comment (modelscope/ms-swift#5611)

my_odps_dataset是文件夹还是路径

如果是文件夹，你试试ms_dataset_id='xxx'

**Comment by Jintao-Huang** [comment]
你可以试试修改DatasetMeta中的load_function

**Comment by onex7777** [comment]
load_function这个函数就是我自定义的函数，目前就是我下shell脚本中传了自定义数据处理脚本py，也传了一个人自定义数据处理脚本中定义的dataset＿id（这个就是 my＿odps＿dataset），但是模型会默认为我这个是modelscope或者hf的数据（它会去自动下载，而我是自己数据库拉下来的数据进过自定义数据处理脚本odps＿dataset.py处理后自定义的数据id），现解决这个不下载而是使用自定义脚本处理的数据？



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年08月31日 23:48 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [modelscope/ms-swift] 从数据库里取sql表数据后，怎样通过自定义处理数据脚本处理后，在训练时shell文件中配置参数 (Issue #5611) |
Jintao-Huang left a comment (modelscope/ms-swift#5611)

你可以试试修改DatasetMeta中的load_function
